src/descent/dogleg.jl

```diff
@@ -0,0 +1,136 @@
+
+Switch between Newton's method and the steepest descent method depending on the size of the
+trust region. The trust region is specified via keyword argument `trust_region` to
+`solve!`.
+
+See also [`SteepestDescent`](@ref), [`NewtonDescent`](@ref), [`DampedNewtonDescent`](@ref).
+"""
+@concrete struct Dogleg <: AbstractDescentAlgorithm
+    newton_descent
+    steepest_descent
+end
+
+function Base.show(io::IO, d::Dogleg)
+    print(io,
+        "Dogleg(newton_descent = $(d.newton_descent), steepest_descent = $(d.steepest_descent))")
+end
+
+supports_trust_region(::Dogleg) = true
+get_linear_solver(alg::Dogleg) = get_linear_solver(alg.newton_descent)
+
+function Dogleg(; linsolve = nothing, precs = DEFAULT_PRECS, damping = False,
+        damping_fn = missing, initial_damping = missing, kwargs...)
+    if damping === False
+        return Dogleg(NewtonDescent(; linsolve, precs), SteepestDescent(; linsolve, precs))
+    end
+    if damping_fn === missing || initial_damping === missing
+        throw(ArgumentError("`damping_fn` and `initial_damping` must be supplied if \
+                             `damping = Val(true)`."))
+    end
+    return Dogleg(DampedNewtonDescent(; linsolve, precs, damping_fn, initial_damping),
+        SteepestDescent(; linsolve, precs))
+end
+
+@concrete mutable struct DoglegCache{pre_inverted, normalform} <:
+                         AbstractDescentCache
+    δu
+    δus
+    newton_cache
+    cauchy_cache
+    internalnorm
+    JᵀJ_cache
+    δu_cache_1
+    δu_cache_2
+    δu_cache_mul
+end
+
+@internal_caches DoglegCache :newton_cache :cauchy_cache
+
+function SciMLBase.init(prob::AbstractNonlinearProblem, alg::Dogleg, J, fu, u;
+        pre_inverted::Val{INV} = False, linsolve_kwargs = (;), abstol = nothing,
+        reltol = nothing, internalnorm::F = DEFAULT_NORM, shared::Val{N} = Val(1),
+        kwargs...) where {F, INV, N}
+    newton_cache = SciMLBase.init(prob, alg.newton_descent, J, fu, u; pre_inverted,
+        linsolve_kwargs, abstol, reltol, shared, kwargs...)
+    cauchy_cache = SciMLBase.init(prob, alg.steepest_descent, J, fu, u; pre_inverted,
+        linsolve_kwargs, abstol, reltol, shared, kwargs...)
+    @bb δu = similar(u)
+    δus = N ≤ 1 ? nothing : map(2:N) do i
+        @bb δu_ = similar(u)
+    end
+    @bb δu_cache_1 = similar(u)
+    @bb δu_cache_2 = similar(u)
+    @bb δu_cache_mul = similar(u)
+
+    T = promote_type(eltype(u), eltype(fu))
+
+    normal_form = prob isa NonlinearLeastSquaresProblem &&
+                  __needs_square_A(alg.newton_descent.linsolve, u)
+    JᵀJ_cache = !normal_form ? J * _vec(δu) : nothing  # TODO: Rename
+
+    return DoglegCache{INV, normal_form}(δu, δus, newton_cache, cauchy_cache, internalnorm,
+        JᵀJ_cache, δu_cache_1, δu_cache_2, δu_cache_mul)
+end
+
+# If TrustRegion is not specified, then use a Gauss-Newton step
+function SciMLBase.solve!(cache::DoglegCache{INV, NF}, J, fu, u, idx::Val{N} = Val(1);
```

Comment:
I don't know about putting these all on the main init and solve because they are meant to be internal?

Reply:
makes sense. dispatching on `__internal_init` and `__internal_solve!`